src/main/java/org/lantern/event/PublicIpAndTokenTracker.java

```diff
@@ -10,58 +11,52 @@
 import com.google.inject.Singleton;
 
 /**
- * This class centralizes tracking of the combination of a valid refresh 
- * token and a valid proxy. Several services rely on that combination,
- * including the XMPP server login and connecting to the friends API.
+ * This class centralizes tracking of the combination of a valid refresh token
+ * and a valid proxy. Several services rely on that combination, including the
+ * XMPP server login and connecting to the friends API.
  */
 @Singleton
 public class PublicIpAndTokenTracker {
 
     private final Logger log = LoggerFactory.getLogger(getClass());
-    
-    private boolean hasRefresh = false;
 
-    private boolean hasPublicIp = false;
+    private volatile boolean hasPublicIp = false;
+    private volatile String refresh;
 
-    private String refresh;
-    
     @Inject
     public PublicIpAndTokenTracker(final Model model) {
         log.debug("Creating tracker");
         Events.register(this);
         this.refresh = model.getSettings().getRefreshToken();
-        if (StringUtils.isNotBlank(refresh)) {
-            this.hasRefresh = true;
-        }
-    }
-    
-    /**
-     * Resets the state of the tracker (useful when reinitializing system).
-     */
-    public void reset() {
-        this.hasPublicIp = false;
     }
     
     @Subscribe
-    public void onRefreshToken(final RefreshTokenEvent refreshEvent) {
-        log.debug("Got refresh token -- loading friends");
-        this.refresh = refreshEvent.getRefreshToken();
-        synchronized (this) {
-            this.hasRefresh = true;
-            if (this.hasPublicIp) {
-                Events.asyncEventBus().post(new PublicIpAndTokenEvent(this.refresh));
-            }
+    synchronized public void onPublicIp(final PublicIpEvent pce) {
+        log.debug("Got public IP");
+        this.hasPublicIp = true;
+        if (StringUtils.isNotBlank(refresh)) {
+            log.debug("Have both public IP and refresh token, posting event");
+            Events.asyncEventBus()
+                    .post(new PublicIpAndTokenEvent(this.refresh));
         }
     }
-    
+
     @Subscribe
-    public void onPublicIp(final PublicIpEvent pce) {
-        synchronized (this) {
-            this.hasPublicIp = true;
-            if (this.hasRefresh) {
-                log.debug("Posting event!!");
-                Events.asyncEventBus().post(new PublicIpAndTokenEvent(this.refresh));
-            }
+    synchronized public void onConnectivityChanged(ConnectivityChangedEvent cce) {
```

Comment:
So we actually have the Events.inOrderAsyncEventBus for making sure events are in order, which we should use if we need that. Beyond that, though, I wonder if there's an associated race condition in PublicIpInfoHandler. In particular, in seems like the only way in practice the public IP should get delivered before the reset is if it's cached. Any connectivitychangedevent should clear the cache, which it does, but it's not thread safe.